doc/designs/ech-api.md

```diff
@@ -0,0 +1,590 @@
+                                  unsigned int protos_len);
+int SSL_CTX_ech_set_outer_alpn_protos(SSL_CTX *s, const unsigned char *protos,
+                                      unsigned int protos_len);
+```
+
+If a client attempts ECH but that fails, or sends an ECH-GREASEd CH, to
+an ECH-supporting server, then that server may return an ECH "retry-config"
+value that the client could choose to use in a subsequent connection. The
+client can detect this situation via the ``SSL_ech_get_status()`` API and
+can access the retry config value via:
+
+```c
+int SSL_ech_get_retry_config(SSL *s, unsigned char **ec, size_t *eclen);
+```
+
+Clients that need fine control over which ECHConfig (from those available) will
+be used, can query the SSL connection, retrieving information about the set of
+ECHConfig values available, and then, if desired, down-select to one of those,
+e.g., based on the ``public_name`` that will be used. This would enable a
+client that selects the server address to use based on IP address hints that
+can also be present in an HTTPS/SCVB resource record to ensure that the correct
+matching ECHConfig is used. The information is presented to the client using
+the ``OSSL_ECH_INFO`` type, which provides a simplified view of ECHConfig data,
+but where each element of an array corresponds to exactly one ECH public value
+and set of names.
+
+```c
+/*
+ * Application-visible form of ECH information from the DNS, from config
+ * files, or from earlier API calls. APIs produce/process an array of these.
+ */
+typedef struct ossl_ech_info_st {
+    int index; /* externally re-usable reference to this value */
+    char *public_name; /* public_name from API or ECHConfig */
+    char *inner_name; /* server-name (for inner CH if doing ECH) */
+    char *outer_alpns; /* outer ALPN string */
+    char *inner_alpns; /* inner ALPN string */
+    char *echconfig; /* a JSON-like version of the associated ECHConfig */
+} OSSL_ECH_INFO;
+
+void OSSL_ECH_INFO_free(OSSL_ECH_INFO *info, int count);
+int OSSL_ECH_INFO_print(BIO *out, OSSL_ECH_INFO *info, int count);
+int SSL_ech_get_info(SSL *s, OSSL_ECH_INFO **info, int *count);
+int SSL_ech_reduce(SSL *s, int index);
+```
+
+The ``SSL_ech_reduce()`` function allows the caller to reduce the active set of
+ECHConfig values down to just the one they prefer, based on the
+``OSSL_ECH_INFO`` index value and whatever criteria the caller uses to prefer
+one ECHConfig over another (e.g. the ``public_name``).
+
+If a client wishes to GREASE ECH using a specific HPKE suite or ECH version
+(represented by the TLS extension type code-point) then it can set those values
+via:
+
+```c
+int SSL_ech_set_grease_suite(SSL *s, const char *suite);
+int SSL_ech_set_grease_type(SSL *s, uint16_t type);
+```
+
+ECH Status API
+--------------
+
+Clients and servers can check the status of ECH processing
+on an SSL connection using this API:
+
+```c
+int SSL_ech_get_status(SSL *s, char **inner_sni, char **outer_sni);
```

Comment:
This seems like kind of a `get1`

Reply:
I'm never sure of the get0 vs. get1 thing tbh but fine with making that change if it's correct.